plutarch-orphanage/src/PlutusLedgerApi/V2/Orphans.hs

```diff
@@ -0,0 +1,1271 @@
+    Builtins.matchData
+      dat
+      shrinkConstr
+      shrinkMap
+      shrinkList
+      (fmap (Builtins.mkI . getNonNegative) . shrink . NonNegative)
+      (fmap Builtins.mkB . shrink)
+    where
+      shrinkConstr :: Integer -> [PlutusTx.BuiltinData] -> [PlutusTx.BuiltinData]
+      shrinkConstr ix dats = do
+        NonNegative ix' <- shrink (NonNegative ix)
+        dats' <- shrink dats
+        pure . Builtins.mkConstr ix' $ dats'
+      shrinkMap :: [(PlutusTx.BuiltinData, PlutusTx.BuiltinData)] -> [PlutusTx.BuiltinData]
+      shrinkMap kvs = Builtins.mkMap <$> shrink kvs
+      shrinkList :: [PlutusTx.BuiltinData] -> [PlutusTx.BuiltinData]
+      shrinkList ell = Builtins.mkList <$> shrink ell
+
+-- | @since 1.0.0
+instance CoArbitrary PlutusTx.BuiltinData where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary dat =
+    Builtins.matchData
+      dat
+      (\ix dats -> variant (0 :: Int) . coarbitrary ix . coarbitrary dats)
+      (\kvs -> variant (1 :: Int) . coarbitrary kvs)
+      (\ell -> variant (2 :: Int) . coarbitrary ell)
+      (\i -> variant (3 :: Int) . coarbitrary i)
+      (\bs -> variant (4 :: Int) . coarbitrary bs)
+
+-- | @since 1.0.0
+instance Function PlutusTx.BuiltinData where
+  {-# INLINEABLE function #-}
+  function = functionMap into outOf
+    where
+      into ::
+        PlutusTx.BuiltinData ->
+        Either
+          (Integer, [PlutusTx.BuiltinData])
+          ( Either
+              [(PlutusTx.BuiltinData, PlutusTx.BuiltinData)]
+              ( Either
+                  [PlutusTx.BuiltinData]
+                  ( Either Integer PlutusTx.BuiltinByteString
+                  )
+              )
+          )
+      into dat =
+        Builtins.matchData
+          dat
+          (\ix -> Left . (ix,))
+          (Right . Left)
+          (Right . Right . Left)
+          (Right . Right . Right . Left)
+          (Right . Right . Right . Right)
+      outOf ::
+        Either
+          (Integer, [PlutusTx.BuiltinData])
+          ( Either
+              [(PlutusTx.BuiltinData, PlutusTx.BuiltinData)]
+              ( Either
+                  [PlutusTx.BuiltinData]
+                  ( Either Integer PlutusTx.BuiltinByteString
+                  )
+              )
+          ) ->
+        PlutusTx.BuiltinData
+      outOf = \case
+        Left (ix, dats) -> Builtins.mkConstr ix dats
+        Right (Left kvs) -> Builtins.mkMap kvs
+        Right (Right (Left ell)) -> Builtins.mkList ell
+        Right (Right (Right (Left i))) -> Builtins.mkI i
+        Right (Right (Right (Right bs))) -> Builtins.mkB bs
+
+-- | @since 1.0.0
+deriving via PlutusTx.BuiltinData instance Arbitrary PLA.Datum
+
+-- | @since 1.0.0
+deriving via PlutusTx.BuiltinData instance CoArbitrary PLA.Datum
+
+-- | @since 1.0.0
+instance Function PLA.Datum where
+  {-# INLINEABLE function #-}
+  function = functionMap coerce PLA.Datum
+
+-- Note: This works because TxId and DatumHash are the same hash type, with the
+-- same length.
```

Comment:
I think it'd be better to have definition explicitly here. For the sake of clarity.

Reply:
Sure, it is a bit cryptic as it stands.

Reply:
Consider having a Blake2b256Hash from which these instances derive? Perhaps that clarifies things better